created basket-service, added swagger, security and exception handler

GrigoriiEgorov opened the pull request:
В данном ПР содержится:
1. Сздан basket-service
2. Добавление "Exception handler" 
3.  Внедрена security на методы 
4.  Добавленна swagger annotation

GrigoriiEgorov committed:
first push

DameerGamlet commented:
Хотелось бы видеть, что было сделано в данном ПР. 

Например,
"Был создан новый микросервис basket-service для работы с корзиной пользователей. Также BasketController, который может добавлять товар в корзину (и т. д.). 

Каждый метод покрыт тестами (интегральными, модульными и т.п.). 

Настроен flyway, и т.д.

Весь сервис доступен и готов к запуску, а также был добавлен автоматический деплой. "

DameerGamlet commented:
Сорри, случайно закрыл ПР.

Я вижу очень много изменений, то название такое "first push" не верное. 

Лучше будет, например, "Adding a "shopping cart" service with minimal operations" или же на русском "Добавление сервиса "корзина" с минимальными операциями". 

Название должно быть осмысленным и не очень длинным. 
В названии никакой избыточной информации (Например, ... с интегральными тестами на все методы получения и бла бла бла). Тут возможно достаточно просто "Добавление нового сервсиа работы с корзиной" и что-то похожее на это. 
Не должны использоваться глаголы в прошедшем времени. Например "Добавил сервис" и т.д., потому что это не описывает то, что сделано в ПР.

Правил на самом деле много разных, так правила наименования для багов, то "Решения бага #1234".

DameerGamlet commented on basket-service/src/main/java/com/yandex/market/basketservice/controller/BasketController.java:
```diff
@@ -0,0 +1,53 @@
+    @GetMapping("/{userId}/basket/products")
+    @ResponseStatus(OK)
+    public Page<ProductResponseDto> getAllProducts(@PathVariable UUID userId,
+                                                   @PageableDefault(size = 10,sort = "id", direction = Sort.Direction.DESC)Pageable pageable) {
```
Тут не хватает пробелов в некоторых места. Это, конечно, решается одним сочитанием клавиш в IDEA, но хотелось бы чтобы форматирование кода было.

askfarz replied:
Согласен, если вдруг не знаешь, то это ctrl + L, либо при коммите там можно галочки поставить чтоб код редактировался на эти моменты. Лучше везде придерживаться одного стиля с пробелами между названием класса и методами, между переменными. Установи еще плагин Sonar

DameerGamlet commented on basket-service/src/main/java/com/yandex/market/basketservice/dto/ProductRequestDto.java:
```diff
@@ -0,0 +1,14 @@
+import jakarta.validation.constraints.Positive;
+
+import java.util.UUID;
+
```
Предлагаю меред аннотациями ставить пустую строку. Это улучшает читаемость в разы.
```java
public record ProductRequest(
        UUID productId,

        @NotNull(message = "The number of items must be specified")
        @Positive(message = "The number of items must be above zero")
        Integer numberOfItems
) {
}
```
Но если всем норм и так, то лучше резолвить коммент.

DameerGamlet commented on basket-service/src/main/java/com/yandex/market/basketservice/mapper/BasketResponseMapper.java:
```diff
@@ -0,0 +1,32 @@
+import java.util.stream.Collectors;
+
+@Component
+@RequiredArgsConstructor
```
BasketMapper мне кажется будет достаточно.
При добавлении корзины возможно будет преобразование с BasketRequest в Basket чтобы запись добавить в БД (ну или возможно будет в ProductMapper).

DameerGamlet commented on basket-service/src/main/java/com/yandex/market/basketservice/model/Basket2Product.java:
```diff
@@ -0,0 +1,30 @@
+@NoArgsConstructor
+@AllArgsConstructor
+@Table(name = "basket_product")
+public class Basket2Product {
```
Извините, но зачем 2 в названии, если класс всё равно называется basket_product без каких либо двоек.

GrigoriiEgorov replied:
2 как to, предлог, показывающий отношение. Также по названию класса сразу понятно, что это Entity описывает связующую таблицу для реализации ManyToMany. На самом деле это не долго поменять, если это не соответсвует общей стилистики.

DameerGamlet replied:
Хорошо. Можно резолвить)

DameerGamlet commented on basket-service/src/main/java/com/yandex/market/basketservice/model/Product.java:
```diff
@@ -0,0 +1,30 @@
+
+    private UUID externalId;
+
+    private Integer availableAmountOfItems;
```
Поле кажется слишком гпомозкин. Мне кажется достаточно будет назвать: itemCount или же numberOfItems.

GrigoriiEgorov replied:
убрал вовсе это поле, как я понял, эта информация храниться в сервисе продукта

DameerGamlet commented on basket-service/src/main/java/com/yandex/market/basketservice/service/BasketService.java:
```diff
@@ -0,0 +1,110 @@
+    }
+
+    public Integer changeNumberOfProductsInBasket(UUID userId, ProductRequestDto productRequestDto) throws Exception {
+        val product = productRepository
```
Мне кажется это можно оптимизировать.
В данный момент:
1) Стучимся к БД и ищем товар.
2) Проверяем что товар есть.
3) Стучимся к БД и Ищём корзину для пользователя.
4) Стучимся к БД и ищёт товар из отношения "карзина - товар".  Если товар есть, то создаём новое отношение корзина-товар. 
5) Устанавливаем новое значение
6) Сохраняем
7) Возрашаем (я не до конца понимаю что возращаем).

DameerGamlet replied:
Итого >= 3 обращения к бд. Надо будет всем вместе подумать как это можно будет оптимизировать. Точно правильное решение без "трогания кода вручную" дать не могу, но могу предположить что что-то из этого можно решить соединениями, джоинами или другими соединениями и отношениями. То есть найти корзину, объеденить  со всеми отношениями корзина-продукт и там же проверить что товар есть и вернуть basketProduct. Если не вернётся  basketProduct, то можно будет предполагать что нет либо корзины, либо товара ну или что-то ещё.

Тут на первый взгляд сложно сказать однозначно.

DameerGamlet commented on basket-service/src/main/java/com/yandex/market/basketservice/service/BasketService.java:
```diff
@@ -0,0 +1,110 @@
+                            product.getExternalId()
+                    ));
+        }
+
```
Как вы относитесь к использованию val и var? Я вот читаю, читаю и явно не понимается что возращается и что отдаётся.
Я заглянул и репозиторий и увидел Optional<Basket2Product>. 
Тут Basket2Product, всё понятно, но дальше возращается return basket.getBasket2Products().size() + 1; и так как  basket также типа var, то что он из себя представляет также сложно понять.

Я пониманию использование в виде var PI = 3.14, или var hello = "Hello", но тут это усложняет пониманию и чтение кода тех, кто с этим методом не работает и будет работать.

Хотел бы услышать мнение всех, кто читает этот коммент.

GrigoriiEgorov replied:
действительно непонятно. В idea все замечательно подписывается, что возвращается... но если читать здесь, то становится неочевидным

GrigoriiEgorov replied:
постарался var  и val использовать только там, где очевидно, что возвращает метод

DameerGamlet replied:
Спасибо

DameerGamlet commented on basket-service/src/main/java/com/yandex/market/basketservice/service/BasketService.java:
```diff
@@ -0,0 +1,110 @@
+        return basket.getBasket2Products().size() + 1;
+    }
+
+    public void deleteProductsSet(UUID userId, Set<UUID> productIdsSet) {
```
Вариант кажется немного громозким и тяжёлым.
Могу предложить:
```java
    @Modifying
    @Query(value = """
                UPDATE XXX x
                SET x.isDeleted=true
                WHERE p.userId=:userId AND x.externalId IN :productIdsSet AND
                """)
    void deleteProductsFromBasket(UUID userId, List<UUID> productIdsSet);
```
Это только пример

DameerGamlet replied:
Также делается для полного удаления с DELETE

askfarz commented on basket-service/src/main/java/com/yandex/market/basketservice/controller/BasketController.java:
```diff
@@ -0,0 +1,53 @@
+    @SneakyThrows
+    @PostMapping("/{userId}/basket/product")
+    @ResponseStatus(OK)
+    public Integer changeNumberOfProductsInBasket(@PathVariable UUID userId,
```
У каждого свой конечно стиль, но в проекте обычно пишем аннотации елочкой, от меньшей длины к большей) Так чуть чуть эстетичнее выглядит)

DameerGamlet replied:
Я не согласен.
    @PostMapping("/{userId}/basket/product") я преподчитаю видеть в виде первой аннотации метода ВСЕГДА.

Да, вкусовщина, но мне кажется а данном случае можно (в методах контроллера) это правило опустить.

DameerGamlet replied:
Хотелось бы услышать мнение остальных

GrigoriiEgorov replied:
Согласен с Дамиром

askfarz commented on basket-service/src/main/java/com/yandex/market/basketservice/dto/ProductRequestDto.java:
```diff
@@ -0,0 +1,14 @@
+        UUID productId,
+        @NotNull(message = "The number of items must be specified")
+        @Positive(message = "The number of items must be above zero")
+        Integer numberOfItems
```
Я вот не помню, если в месседже не указать там потом укажется какое поле валидацию не проходит? Если нет, то можем указать имя поле в сообщении

GrigoriiEgorov replied:
"message" обязательно указывать. редактор высвечивает ошибку, если убрать

askfarz commented on basket-service/src/main/java/com/yandex/market/basketservice/model/Basket.java:
```diff
@@ -0,0 +1,28 @@
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.SEQUENCE)
+    private Long id;
```
Не забудь добавить еще @SequenceGenerator

askfarz commented on basket-service/src/main/java/com/yandex/market/basketservice/model/Basket2Product.java:
```diff
@@ -0,0 +1,30 @@
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.SEQUENCE)
+    private long id;
```
Тут лучше все таки использовать Long, чтобы везде id были одинаковые и если нужно будет работать черех интерфейс JPA

askfarz commented on basket-service/src/main/java/com/yandex/market/basketservice/repository/BasketProductRepository.java:
```diff
@@ -0,0 +1,20 @@
+import java.util.Optional;
+
+public interface BasketProductRepository extends JpaRepository<Basket2Product, Long> {
+    @Query(value = "SELECT * FROM basket_product WHERE basket_id=:basket_id AND product_id=:product_id", nativeQuery = true)
```
тесно :(

GrigoriiEgorov committed:
* added url in BasketController.java;
* added handleNotEnoughIGoodsInStockException in BasketExceptionHandler.java;
* deleted:
  - BasketResponseDto.java,
  - ProductResponseDto.java,
  - NotEnoughIGoodsInStockException.java,
  - BasketResponseMapper.java,
  - ProductResponseMapper.java,
  - BasketProductRepository.java,
  - ProductRepository.java;
* changed 'String statusCode' to 'UUID id' in ErrorResponse.java;
* added interface ItemResponse.java;
* changed Basket.java;
* renamed Basket2Product.java to BasketItem.java;
* added BasketItemId.java;
* renamed Product.java to Item.java;
* changed ItemRepository.java;
* changed BasketService.java all methods;
* added ExceptionMessagesConstants.PRODUCT_LIST_FOR_REMOVE_IS_NULL;

GrigoriiEgorov committed:
Merge branch 'develop' into feature/create-basket-service

GrigoriiEgorov committed:
* adding application.yaml
* correction ItemRepository.java

GrigoriiEgorov committed:
Merge branch 'develop' into feature/create-basket-service

GrigoriiEgorov committed:
Merge branch 'develop' into feature/create-basket-service

GrigoriiEgorov committed:
* adding new methods in BasketExceptionHandler.java;
* delete ErrorResponse.java;
* adding validate annotation in ItemRequest.java
* changing parameters for @EqualsAndHashCode in Basket.java;
* adding validation in BasketService.java

GrigoriiEgorov committed:
* adding swagger annotation

askfarz commented on basket-service/build.gradle:
```diff
@@ -1,39 +1,65 @@
+//
+//tasks.named('test') {
+//    useJUnitPlatform()
+//}
```
Удаляй сразу все ненужное, не оставляй так комментами)

DameerGamlet commented on basket-service/src/main/java/com/yandex/market/basketservice/controller/BasketApi.java:
```diff
@@ -0,0 +1,122 @@
+})
+public interface BasketApi {
+
+    @Operation(
```
Я понимаю что логи хорошие, но мы решили сваггерские логи писать на русском

DameerGamlet replied:
Это было (должно было быть) реализовано везде.
Логи не сваггерские - можно на английском

DameerGamlet commented on basket-service/src/main/java/com/yandex/market/basketservice/controller/BasketApi.java:
```diff
@@ -0,0 +1,122 @@
+    )
+    Integer deleteItemsList(
+            @Parameter(name = "id", description = "user external id", required = true) UUID userId,
+            @Parameter(name = "itemsIds", description = "list items for deleting", required = true) List<UUID> itemsIdsList
```
itemsIdsList - лучше будет itemIds (без List и item в единственном числе)

DameerGamlet replied:
Также будет:
 ItemRequest itemRequest изменён на  ItemRequest request

DameerGamlet commented on basket-service/src/main/java/com/yandex/market/basketservice/controller/BasketController.java:
```diff
@@ -0,0 +1,52 @@
+        return basketService.getAllItemsInsideBasketByUserId(userId, pageable);
+    }
+
+    @PatchMapping("/users/{userId}/basket/product")
```
```suggestion
    @PatchMapping("/users/{userId}/basket/products")
```

askfarz commented on basket-service/src/main/java/com/yandex/market/basketservice/config/SecurityConfig.java:
```diff
@@ -0,0 +1,51 @@
+//    public WebSecurityCustomizer webSecurityCustomizer() {
+//        return web -> web.ignoring()
+//                .requestMatchers("/public/api/v1/product-previews");
+//    }
```
Это можно удалить

askfarz commented on basket-service/src/main/java/com/yandex/market/basketservice/controller/BasketExceptionHandler.java:
```diff
@@ -0,0 +1,137 @@
+    @ExceptionHandler(ConstraintViolationException.class)
+    @ResponseStatus(HttpStatus.BAD_REQUEST)
+    public ErrorResponse handleConstraintViolationException(ConstraintViolationException ex) {
+        String errorId = UUID.randomUUID().toString();
```
А зачем давать айдишник для ошибки? Мы же его нигде не храним и по айдишники вряд ли сможем определить как-то. Или как?

GrigoriiEgorov replied:
зацел на будущее. Кажется, что такие как ошибки нужо писать и хранить

askfarz commented on basket-service/src/main/java/com/yandex/market/basketservice/controller/BasketExceptionHandler.java:
```diff
@@ -0,0 +1,137 @@
+//                ex.getMessage(),
+//                OffsetDateTime.now()
+//        );
+//    }
```
Что с этим?

askfarz commented on basket-service/src/main/java/com/yandex/market/basketservice/model/BasketItem.java:
```diff
@@ -0,0 +1,39 @@
+@NoArgsConstructor
+@EqualsAndHashCode(of = {"basket", "item"})
+@Table(name = "basket_item")
+public class BasketItem {
```
@Table(name = "basket_items")
Названия таблиц во множ числе

GrigoriiEgorov replied:
в таком случае, кажется правильнее будет baskets_items

DameerGamlet reviewed (commented):
Резолви после коммита

GrigoriiEgorov committed:
resolved some comments
